### ISSUE TITLE
Compiler does not resolve implicit type class instance

### ISSUE THREAD

**Issue by justjoheinz** [idris-lang/Idris-dev#2771]
In an attempt to write a Unit testing framework I started with:

```idris
data AssertType = Equals | NotEquals

interpAssertType: AssertType ->  Type
interpAssertType Equals  =  {a : Type} -> (eq: Eq a) => a -> a -> Bool
interpAssertType NotEquals = {a : Type} -> (eq: Eq a) => a -> a -> Bool

toFunctionAssertType: (at: AssertType) -> (interpAssertType at)
toFunctionAssertType Equals = (==)
toFunctionAssertType NotEquals = (/=)

assert: (at: AssertType) -> (interpAssertType at)
assert x = toFunctionAssertType x
```

This leads to the following REPL sequence:
```
*UnitTest> :t assert
assert : (at : AssertType) -> interpAssertType at
*UnitTest> :t assert Equals
assert Equals : Eq a => a -> a -> Bool
*UnitTest> assert Equals "Hello" "world"
(input):Incomplete term ([__])
*UnitTest> :let a = assert Equals
*UnitTest> a "hello" "world"
False : Bool
```

Attempts to get this fixed on the IRC channel failed. If this is a bug or something else is also a bit unclear :-)

**Comment by bjz** [comment]
Yes! I get this error with https://gist.github.com/bjz/01d39a8b73da9bc7138e/74eae0c4ba4b16a94345d563a424b76310fb91d0:

```
*Fmt> format "hello {} hi" "howdy"
(input):Incomplete term ([__])
```

Not sure if it is related!

**Comment by bjz** [comment]
Interesting!

```
*Fmt> :let f = format "hello {}"
*Fmt> f 1
"hello 1" : String
```

**Comment by justjoheinz** [comment]
It would be nice to get a comment from someone in the know about this issue.

**Comment by justjoheinz** [comment]
Still present with Idris 0.11

**Comment by ahmadsalim** [comment]
You can work around this by using the following:

```idris
data AssertType = Equals | NotEquals

toFunctionAssertType: (Eq a) => (at: AssertType) -> a -> a -> Bool
toFunctionAssertType Equals = (==)
toFunctionAssertType NotEquals = (/=)

assert: (Eq a) => (at: AssertType) -> a -> a -> Bool
assert x = toFunctionAssertType x
```

I think there are some issues with scoped implicits, but I am not knowledgeable enough in general to tell you why the resolution does not happen correctly.

**Comment by justjoheinz** [comment]
But what if I do not want to promote the equality to the `toFunctionAssertType`? The next logical conclusion would be to write a SMALLER case which would be along the line of:

```idris
data AssertType = Equals | NotEquals | Smaller
...
toFunctionAssertType: (at: AssertType) -> (interpAssertType at)
...
interpAssertType Smaller = {a : Type} -> (sm: Ord a) => a -> a -> Bool
``` 
I have to admit that I am not to familiar with idris at the moment, so another workaround would be welcome, @ahmadsalim.

**Comment by ahmadsalim** [comment]
@justjoheinz I haven't tested i yet but I believe you could put the constraints in GADTs and it will resolve them for you, e.g.:

data AssertType : (a : Type) -> Type where
     Equals : (Eq a) => AssertType a
     NotEquals : (Eq a) => AssertType a
     Smaller : (Ord a) => AssertType a

toFunctionAssertType: (at: AssertType a) -> a -> a -> Bool
toFunctionAssertType (Equals @{eq}) = (==) @{eq}
toFunctionAssertType (NotEquals @{eq}) = (/=) @{eq}
toFunctionAssertType (Smaller @{ord}) = (<=) @{ord}

assert: (at: AssertType a) -> a -> a -> Bool
assert x = toFunctionAssertType x

**Comment by ahmadsalim** [comment]
Actually, the following seems to work:

```idris
data AssertType : (a : Type) -> Type where
     Equals : (Eq a) => AssertType a
     NotEquals : (Eq a) => AssertType a
     Smaller : (Ord a) => AssertType a

toFunctionAssertType: (at: AssertType a) -> a -> a -> Bool
toFunctionAssertType Equals = (==)
toFunctionAssertType NotEquals = (/=)
toFunctionAssertType Smaller = (<=)

assert: (at: AssertType a) -> a -> a -> Bool
assert x = toFunctionAssertType x
```

**Comment by justjoheinz** [comment]
@ahmadsalim Could you still clarify if what I wrote initially would be supposed to work? Or is the workaround a workaround against a compiler error or a non fixable compiler limitation? Is this actually a bug?

**Comment by ahmadsalim** [comment]
@justjoheinz I do not have enough expertise to judge whether this should work or not.

Personally, I don't think that your code looks unreasonable from first sight, but it might be non-trivial to support such kind of features.

**Comment by justjoheinz** [comment]
Due to this sentence in the CHANGELOG I was hopeful, that something might have changed:
Constraint arguments can now appear anywhere in function types, not just at the top level or after an implicit argument binding.

However the error (if it is one) persists in idris 0.12. I would be really interested in knowing if my first attempt to solve my problem is actually valid (because all the workarounds have severe limitations beyond application of my small example snippet).
